File: tests/cli/integration/help-version.test.ts
import { spawn } from "node:child_process";
import { promises as fs } from "node:fs";
import * as os from "node:os";
import * as path from "node:path";

import { describe, expect, it } from "vitest";

import { BIN_PATH, ensureBuilt, runCli } from "./spawn.js";

describe("--help / --version (AC-20)", () => {
  it("--version prints package version + exits 0", async () => {
    await ensureBuilt();
    const r = await runCli({ args: ["--version"] });
    expect(r.exitCode).toBe(0);
    expect(r.stdout).toMatch(/claude-profiles \d+\.\d+\.\d+/);
  });

  // Regression: isDirect must canonicalise argv[1] before comparing it to
  // import.meta.url. Without realpathSync, a symlinked bin path (npm's
  // node_modules/.bin shim, macOS /var→/private/var, Homebrew prefixes,
  // /usr/local/bin → /opt/homebrew/...) compares unequal so main() never
  // runs and the binary silently exits 0 with no output. Found via
  // post-publish install validation of 0.2.3.
  it("--version works when invoked through a symlink (real npm install path)", async () => {
    await ensureBuilt();
    const tmp = await fs.mkdtemp(path.join(os.tmpdir(), "ccp-symlink-bin-"));
    try {
      const linked = path.join(tmp, "claude-profiles");
      await fs.symlink(BIN_PATH, linked);
      const r = await new Promise<{ stdout: string; stderr: string; code: number }>((resolve, reject) => {
        const child = spawn(process.execPath, [linked, "--version"], { stdio: ["ignore", "pipe", "pipe"] });
        let stdout = "";
        let stderr = "";
        child.stdout.on("data", (d) => (stdout += d.toString()));
        child.stderr.on("data", (d) => (stderr += d.toString()));
        child.on("close", (code) => resolve({ stdout, stderr, code: code ?? -1 }));
        child.on("error", reject);
      });
      expect(r.code, `stderr: ${r.stderr}`).toBe(0);
      expect(r.stdout).toMatch(/claude-profiles \d+\.\d+\.\d+/);
    } finally {
      await fs.rm(tmp, { recursive: true, force: true });
    }
  });

  it("-V short form also works", async () => {
    await ensureBuilt();
    const r = await runCli({ args: ["-V"] });
    expect(r.exitCode).toBe(0);
    expect(r.stdout).toMatch(/claude-profiles \d+\.\d+\.\d+/);
  });

  it("--help prints usage with all R29 verbs", async () => {
    await ensureBuilt();
    const r = await runCli({ args: ["--help"] });
    expect(r.exitCode).toBe(0);
    for (const verb of ["init", "list", "use", "status", "drift", "diff", "new", "validate", "sync", "hook"]) {
      expect(r.stdout).toContain(verb);
    }
    expect(r.stdout).toContain("EXIT CODES");
  });

  it("use --help prints verb-specific guidance", async () => {
    await ensureBuilt();
    const r = await runCli({ args: ["use", "--help"] });
    expect(r.exitCode).toBe(0);
    expect(r.stdout).toContain("--on-drift=");
  });

  // claude-code-profiles-xd2: every verb's help follows the same template —
  // tagline, USAGE, DESCRIPTION, EXAMPLES, EXIT CODES. Lock that in so a
  // future help-text edit can't quietly drop a section.
  it("every verb's help has the standard sections", async () => {
    await ensureBuilt();
    const verbs = ["init", "list", "use", "status", "drift", "diff", "new", "validate", "sync", "hook", "doctor", "completions"];
    for (const verb of verbs) {
      const r = await runCli({ args: [verb, "--help"] });
      expect(r.exitCode, `${verb} --help exited non-zero`).toBe(0);
      expect(r.stdout, `${verb} --help missing USAGE`).toContain("USAGE");
      expect(r.stdout, `${verb} --help missing DESCRIPTION`).toContain("DESCRIPTION");
      expect(r.stdout, `${verb} --help missing EXAMPLES`).toContain("EXAMPLES");
      expect(r.stdout, `${verb} --help missing EXIT CODES`).toContain("EXIT CODES");
      // Every verb mentions --cwd or --json (the common globals) so users
      // running just `claude-profiles <verb> --help` don't have to bounce out
      // to the top-level help to learn about them.
      expect(r.stdout, `${verb} --help missing global option reference`).toMatch(/--(cwd|json)/);
    }
  });

  // cw6.4 followup: validate's exit-codes block must document the R44
  // marker-failure case so users grepping --help know why validate may
  // exit 1 instead of 0/3.
  it("validate --help documents the R44 marker-missing exit-1 case", async () => {
    await ensureBuilt();
    const r = await runCli({ args: ["validate", "--help"] });
    expect(r.exitCode).toBe(0);
    expect(r.stdout).toContain("EXIT CODES");
    expect(r.stdout).toMatch(/missing claude-profiles markers/i);
    expect(r.stdout).toContain("claude-profiles init");
  });

  it("top-level --help defines the spec terms it uses", async () => {
    await ensureBuilt();
    const r = await runCli({ args: ["--help"] });
    expect(r.exitCode).toBe(0);
    expect(r.stdout).toContain("GLOSSARY");
    // The four terms the help text references should each appear in the glossary.
    for (const term of ["profile", "extends", "drift", "materialize"]) {
      expect(r.stdout).toContain(term);
    }
  });
});

File: go/tests/integration/help_version_test.go
package integration_test

import (
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strings"
	"testing"

	"github.com/htxryan/c3p/tests/integration/helpers"
)

// TestVersionPrintsAndExitsZero is the IV translation of TS
// help-version.test.ts → "--version prints package version + exits 0".
// The Go bin emits "c3p X.Y.Z"; under tests the version string is the
// fixed "0.0.0-dev" tag injected by the test harness build (no ldflags).
func TestVersionPrintsAndExitsZero(t *testing.T) {
	helpers.EnsureBuilt(t)
	r := mustRun(t, helpers.SpawnOptions{Args: []string{"--version"}})
	if r.ExitCode != 0 {
		t.Fatalf("--version exit: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if matched, _ := regexp.MatchString(`c3p \S+`, r.Stdout); !matched {
		t.Fatalf("--version stdout missing 'c3p X.Y.Z' pattern: %q", r.Stdout)
	}
}

// TestVersionWorksThroughSymlink is the post-publish regression from TS:
// the bin must work when invoked through a symlink (npm shim path).
// Go binaries don't have the import.meta.url problem TS had, but we still
// pin the behavior so a future packaging change can't regress it.
func TestVersionWorksThroughSymlink(t *testing.T) {
	if runtime.GOOS == "windows" {
		t.Skip("symlink test requires non-Windows or developer-mode admin")
	}
	helpers.EnsureBuilt(t)
	bin := helpers.BinPath(t)
	tmp := t.TempDir()
	linked := filepath.Join(tmp, "c3p-link")
	if err := os.Symlink(bin, linked); err != nil {
		t.Fatalf("symlink: %v", err)
	}

	// Spawn through the symlink directly (helpers.RunCli would resolve to
	// the cached real binary path, which would skip the symlink entirely
	// and miss the regression we want to pin).
	r := runBin(t, linked, []string{"--version"}, nil, "")
	if r.ExitCode != 0 {
		t.Fatalf("symlink --version exit: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if matched, _ := regexp.MatchString(`c3p \S+`, r.Stdout); !matched {
		t.Fatalf("symlink --version stdout missing 'c3p X.Y.Z': %q", r.Stdout)
	}
}

// TestVersionShortFlag covers -V parity with --version.
func TestVersionShortFlag(t *testing.T) {
	helpers.EnsureBuilt(t)
	r := mustRun(t, helpers.SpawnOptions{Args: []string{"-V"}})
	if r.ExitCode != 0 {
		t.Fatalf("-V exit: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if matched, _ := regexp.MatchString(`c3p \S+`, r.Stdout); !matched {
		t.Fatalf("-V stdout missing version pattern: %q", r.Stdout)
	}
}

// TestHelpListsR29Verbs asserts every R29 verb appears in --help, plus
// the EXIT CODES section.
func TestHelpListsR29Verbs(t *testing.T) {
	helpers.EnsureBuilt(t)
	r := mustRun(t, helpers.SpawnOptions{Args: []string{"--help"}})
	if r.ExitCode != 0 {
		t.Fatalf("--help exit: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	for _, verb := range []string{"init", "list", "use", "status", "drift", "diff", "new", "validate", "sync", "hook"} {
		if !strings.Contains(r.Stdout, verb) {
			t.Errorf("--help missing verb %q", verb)
		}
	}
	if !strings.Contains(r.Stdout, "EXIT CODES") {
		t.Errorf("--help missing EXIT CODES section: %q", r.Stdout)
	}
}

// TestUseDashHelpPrintsVerbGuidance asserts "use --help" mentions
// --on-drift= (the verb-specific pivot).
func TestUseDashHelpPrintsVerbGuidance(t *testing.T) {
	helpers.EnsureBuilt(t)
	r := mustRun(t, helpers.SpawnOptions{Args: []string{"use", "--help"}})
	if r.ExitCode != 0 {
		t.Fatalf("use --help exit: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if !strings.Contains(r.Stdout, "--on-drift=") {
		t.Errorf("use --help missing --on-drift= guidance: %q", r.Stdout)
	}
}

// TestEveryVerbHelpHasStandardSections is the cw6 lock-in: every verb's
// help follows the same template (USAGE / DESCRIPTION / EXAMPLES /
// EXIT CODES + a global flag reference).
func TestEveryVerbHelpHasStandardSections(t *testing.T) {
	helpers.EnsureBuilt(t)
	verbs := []string{"init", "list", "use", "status", "drift", "diff", "new", "validate", "sync", "hook", "doctor", "completions"}
	flagRE := regexp.MustCompile(`--(cwd|json)`)
	for _, verb := range verbs {
		t.Run(verb, func(t *testing.T) {
			r := mustRun(t, helpers.SpawnOptions{Args: []string{verb, "--help"}})
			if r.ExitCode != 0 {
				t.Fatalf("%s --help exit: want 0, got %d (stderr=%q)", verb, r.ExitCode, r.Stderr)
			}
			for _, sec := range []string{"USAGE", "DESCRIPTION", "EXAMPLES", "EXIT CODES"} {
				if !strings.Contains(r.Stdout, sec) {
					t.Errorf("%s --help missing %s", verb, sec)
				}
			}
			if !flagRE.MatchString(r.Stdout) {
				t.Errorf("%s --help missing global flag reference (--cwd or --json)", verb)
			}
		})
	}
}

// TestValidateHelpDocumentsR44MarkerCase covers cw6.4: validate --help
// must mention the missing-markers exit-1 case + remediation.
func TestValidateHelpDocumentsR44MarkerCase(t *testing.T) {
	helpers.EnsureBuilt(t)
	r := mustRun(t, helpers.SpawnOptions{Args: []string{"validate", "--help"}})
	if r.ExitCode != 0 {
		t.Fatalf("validate --help exit: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if !strings.Contains(r.Stdout, "EXIT CODES") {
		t.Errorf("validate --help missing EXIT CODES")
	}
	lower := strings.ToLower(r.Stdout)
	if !strings.Contains(lower, "missing c3p markers") {
		t.Errorf("validate --help missing R44 marker-missing description: %q", r.Stdout)
	}
	if !strings.Contains(r.Stdout, "c3p init") {
		t.Errorf("validate --help missing 'c3p init' remediation: %q", r.Stdout)
	}
}

// TestTopLevelHelpDefinesGlossary asserts the spec terms in --help all
// appear in the GLOSSARY block.
func TestTopLevelHelpDefinesGlossary(t *testing.T) {
	helpers.EnsureBuilt(t)
	r := mustRun(t, helpers.SpawnOptions{Args: []string{"--help"}})
	if r.ExitCode != 0 {
		t.Fatalf("--help exit: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if !strings.Contains(r.Stdout, "GLOSSARY") {
		t.Errorf("--help missing GLOSSARY section: %q", r.Stdout)
	}
	for _, term := range []string{"profile", "extends", "drift", "materialize"} {
		if !strings.Contains(r.Stdout, term) {
			t.Errorf("GLOSSARY missing term %q", term)
		}
	}
}
